2.0-PROPOSAL/README.md

```diff
@@ -0,0 +1,885 @@
+logic and coordinate Brigade's underlying subsystems through well-defined
+interfaces.
+
+### Record Storage
+
+The record storage subsystem will provide a synchronous interface for the
+persistence of domain objects such as projects and events. Document-based
+storage is proposed since relatively few relationships exist between domain
+objects and among those relationships that do exist, composition is highly
+favored. i.e. A relational database is unlikely to yield any significant
+advantages and would also introduce object-relational impedance that can
+otherwise be avoided.
+
+Further, after careful consideration of MongoDB, Cassandra, and CouchDB, as well
+as cursory consideration of a few other options, MongoDB is proposed as the
+underlying data store based on several factors:
+
+1. Development concerns:
+    1. Programming model has lowest impedence with proposed domain model
+    1. Expressive query language
+    1. Atomic updates to single documents without the need for transactions 
+    1. Widespread developer familiarity
+1. Operational concerns:
+    1. Deployable in-cluster using existing Helm charts; can be included in
+       default Brigade install
+    1. Deployable in a wide array of configurations to cover both
+       evaluation/development _and_ production 
+    1. Managed instances of MongoDB (or wire-compatible services) readily
+       available from (or in) the leading public clouds
+    1. Demonstrated ability to scale
+1. Permissive (enough) license
+
+N.B.: Criticisms of MongoDB such as [this one from
+Jepsen](http://jepsen.io/analyses/mongodb-4.2.6) were carefully considered and
+determined to be inapplicable with respect to the proposed use. Likewise, known
+[licensing
+constraints](https://www.mongodb.com/licensing/server-side-public-license/faq)
+were also carefully considered and judged inapplicable for the use case.
+
+### Scheduling
+
+The scheduling subsystem will provide a synchronous interface for managing the
+underlying workload execution substrate-- which will be Kubernetes, exclusively.
+Additionally, it will provide an agent to asynchronously handle events by
+orchestrating workload execution on the substrate.
+
+The synchronous interface will effect immediate changes to the substrate
+wherever those changes can be applied in real-time and with minimal resource
+utilization. For instance, event creation or cancellation may occur in
+real-time. Event execution will be deferred to the asynchronous agent.
+
+To bridge the gap between the synchronous and asynchronous components of this
+subsystem, a message bus capable of ensuring reliable transmission of messages
+with at-least-once delivery guarantees will be required.
+
+After careful consideration of Kafka, ActiveMQ, RabbitMQ, and Redis (using the
+[reliable queue](https://redis.io/commands/rpoplpush#pattern-reliable-queue)
+pattern), Kafka is proposed to be used for this purpose based on several
+factors:
+
```

Comment:
From @brandonkal:

> Have you considered NATS?

With finite bandwidth and a very large number of contenders in this field, summary dismissal of some candidates was necessary in order to allow closer examination of others that were not as easily dismissed. (There was a primary, if you will.) NATS didn't receive closer scrutiny because upon initial consideration, it appeared too k8s-centric. You'll note that there's a theme to this proposal of making Brigade _less_ k8s-centric overall.

As a courtesy, I will give NATS a closer look.

Reply:
@brandonkal fyi, I've been investigating NATS. I'm more optimistic about it now than I was initially. From a development perspective, the fair scheduling problem can possibly be solved more easily than with Kafka, although I haven't written any code to prove this _yet_.

From an ops perspective, I am a little more reluctant because I haven't yet found a Helm chart to set up NATS _with streaming support_ nor have I found an option to provision a _fully managed_ NATS in any (or preferably several) of the major public clouds. Ideally, we would want to be able to say "we're not experts on running NATS in production; here is some guidance, but here also are options for running it in production with zero fuss."

Stay tuned.

Reply:
Helm chart for setting up NATS streaming: https://github.com/nats-io/k8s#helm-charts-for-nats

An option to deploy NATS as a managed service to the cloud of your choice: https://synadia.com/ngs/pricing

I plan to try these both today.

Reply:
@brandonkal

Update on Kafka v NATS:

First, thank you for encouraging me to give NATS a closer look. My initial assessment that it was "too k8s centric" was in error and after spending some time getting to know NATS and writing some sample code, I found it an absolute joy to work with. It handled our use case better than anything else I looked at. I'll be frank in saying that _from a developer's perspective_, I now have a strong bias toward NATS.

However, from an operations perspective, I cannot endorse it. NATS can easily account for the "batteries included" story since I found Helm charts that would have worked just fine with a few tweaks, but should anyone decide that they do not care for the burden of hosting / operating NATS at scale in production and wish to provision a managed service instead, the situation is bleak, at present. Synadia's offering looks compelling, but it's the _only_ option for managed NATS that I could find and on top of that, it doesn't support NATS streaming (aka "STAN") which is a feature we absolutely need in order to implement our use case.

I look forward to this situation improving in the future and NATS becoming a more tenable option for use with future projects.

Now, for a surprising turn of events...

Having access, as I do, to folks from Azure's messaging team (responsible for products like Azure Service Bus and Azure Event Hubs; i.e. folks who live and breath messaging) I engaged them for some input. Their advice was that I may have been too quick to dismiss AMQP 1.0 and they offered guidance on using it to satisfy our use case. In the end, programmatically, it ended up being a better fit than either Kafka or NATS.

This left the question of how easily the "batteries included" and "managed service" narratives could be fulfilled on the ops side of the equation. ActiveMQ is probably the most popular and battle-tested OSS AMQP 1.0 implementation. There is no production quality Helm chart for this (though there are some interesting projects that are WIP). So, figuring out how to get Artemis (the next gen ActiveMQ) cranking inside a k8s cluster was non-trivial and cost me several days of effort, but in the end I am confident that it can be done _at least_ well enough to satsify the "batteries included" narrative. For those wishing to use a managed service in production, the story is much simpler, as AMQP is a Standard with a capital S (it's an OASIS standard) and I was easily able to locate numerous options for hosted/managed ActiveMQ and other AMQP-compatible services.

So, recapping this surprising twist-- the suggestion to take a closer look at NATS will be resulting in an amendment to the proposal to propose AMQP instead with ActiveMQ Artemis as the "included batteries."